gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/DeleteWorkDirsActivityImpl.java

```diff
@@ -0,0 +1,224 @@
+      optJobName = Optional.ofNullable(jobState.getJobName());
+
+      SharedResourcesBroker<GobblinScopeTypes> instanceBroker = JobStateUtils.getSharedResourcesBroker(jobState);
+      JobContext jobContext = new JobContext(jobState.getProperties(), log, instanceBroker, null);
+      if (PropertiesUtils.getPropAsBoolean(jobState.getProperties(), ConfigurationKeys.CLEANUP_STAGING_DATA_BY_INITIALIZER, "false")) {
+        //Clean up will be done by initializer.
+        return CleanupResult.createEmpty();
+      }
+      try {
+        if (!canCleanStagingData(jobContext, jobState)) {
+          log.error("Job " + jobState.getJobName() + " has unfinished commit sequences. Will not clean up staging data.");
+          return CleanupResult.createEmpty();
+        }
+      } catch (IOException e) {
+        throw new JobException("Failed to check unfinished commit sequences", e);
+      }
+      Map<String, Long> cleanupMetrics = jobState.getPropAsBoolean(ConfigurationKeys.CLEANUP_STAGING_DATA_PER_TASK, ConfigurationKeys.DEFAULT_CLEANUP_STAGING_DATA_PER_TASK) ?
+          cleanupStagingDataPerTask(jobState, workSpec) : cleanupStagingDataForEntireJob(jobState, resourcesToClean, workSpec);
+
+      return new CleanupResult(cleanupMetrics);
+    } catch (Exception e) {
+      throw ApplicationFailure.newNonRetryableFailureWithCause(
+          String.format("Failed to cleanup temporary folders for job %s", optJobName.orElse(UNDEFINED_JOB_NAME)),
+          IOException.class.toString(),
+          new IOException(e)
+      );
+    }
+  }
+
+  private boolean canCleanStagingData(JobContext jobContext, JobState jobState)
+      throws IOException {
+    return jobContext.getSemantics() != DeliverySemantics.EXACTLY_ONCE || !jobContext.getCommitSequenceStore()
+        .get().exists(jobState.getJobName());
+  }
+
+  // Goes through all the task states and cleans them up individually, there is some overlap with the job level cleanup as the job level cleanup will
+  // also clean up the task level staging data
+  private static Map<String, Long> cleanupStagingDataPerTask(JobState jobState, WUProcessingSpec workSpec) throws IOException {
+    Closer closer = Closer.create();
+    Map<String, ParallelRunner> parallelRunners = Maps.newHashMap();
+    FileSystem fs = Help.loadFileSystem(workSpec);
+    int numThreads = jobState.getPropAsInt(ParallelRunner.PARALLEL_RUNNER_THREADS_KEY, ParallelRunner.DEFAULT_PARALLEL_RUNNER_THREADS);
+    List<TaskState> taskStateList = Help.loadTaskStates(workSpec, fs, jobState, numThreads);
+    Map<String, Long> cleanupMetrics = Maps.newHashMap();
+    try {
+      for (TaskState taskState : taskStateList) {
+        try {
+          cleanupMetrics.putAll(cleanTaskStagingData(taskState, log, closer, parallelRunners));
+        } catch (IOException e) {
+          log.error(String.format("Failed to clean staging data for task %s: %s", taskState.getTaskId(), e), e);
+        }
```

Comment:
is it truly necessary to enumerate all of these task states?  `FileSystem.delete(Path, boolean)` takes a `recursive` param